zluda/src/impl/dark_api.rs

```diff
@@ -62,6 +62,31 @@ impl CudaDarkApi for CudaDarkApiZluda {
         device::primary_ctx_get(pctx, hip_dev).into_cuda()
     }
 
+    unsafe extern "system" fn set_device_flags(
+        dev: cuda_types::CUdevice,
+        flags: c_uint,
+    ) -> CUresult {
+        use hip_runtime_sys::*;
+        let hip_dev = FromCuda::from_cuda(dev);
+        let mut active_dev = 0i32;
+        hipGetDevice(&mut active_dev as _);
+        if hip_dev != active_dev {
+            hipSetDevice(hip_dev);
+        }
+        let res = hipSetDeviceFlags(flags).into_cuda();
+        if hip_dev != active_dev {
+            hipSetDevice(active_dev);
```

Comment:
Same here.

Reply:
Error here should not change the result of the already processed request: res.

Reply:
If `hipSetDevice` fails, set_device_flags will change the default device, but the user may not expect it to change the default device.

Reply:
But we can't do anything with that. Returning error code won't change the device back nor can it indicate that device is changed.

In truth, I suspect that the whole device change code is useless and the function is actually called only on active device like cudaSetDeviceFlags.

Reply:
That's right. But I'm assuming the situation that someone calls it with particular device id and coincidentally `hipSetDevice` fails. If so, it will be better to return error instead of `CUDA_SUCCESS` in order to make the user to know something went wrong.

Reply:
Since it's Dark API, there is no user – it's only called internally by cudart* and there it's often better to do nothing and return `CUDA_SUCCESS` than returning any error code as that will fail everything around it that would otherwise work fine.

If you want implementation that would be stable in all cases than it would be:
`
    unsafe extern "system" fn set_device_flags(
        _dev: cuda_types::CUdevice,
        _flags: c_uint,
    ) -> CUresult {
        CUresult::CUDA_ERROR_PRIMARY_CONTEXT_ACTIVE
    }
` That error is the only safe alternative to CUDA_SUCCESS there.